### PR TITLE
Optimize `BigInt.new(Int::Primitive)`

### PR DESCRIPTION
**Pull request by HertzDevil**
* Uses `LibGMP::SI` and `LibGMP::UI` for range checks, for MPIR compatiblity. This means integers that fit into 64 bits but not 32 bits will now use the fast GMP path on 64-bit Windows. (This applies to _all_ `BigInt` and `BigFloat` methods that call any of GMP's `*_si` or `*_ui` functions. Those other methods will be touched later.)
* Merges the `Int::Signed` and `Int::Unsigned` overloads into one. This one is more subtle, but it allows `Int128`s that fit into `UInt64` but not `Int64` to use the fast path.
* For the slow path, uses GMP's low-level functions to populate the bits directly, without going through `Int#to_s`. This provides a tremendous speedup:
  ```crystal
  require "big"
  require "benchmark"

  struct BigInt
    def self.new(*, num2 num : Int::Primitive)
      # new implementation
    end
  end
  
  x = 1.to_big_i
  
  Benchmark.ips do |b|
    b.report("old") { x = BigInt.new(0x0123456789abcdef_i64) }
    b.report("new") { x = BigInt.new(num2: 0x0123456789abcdef_i64) }
  end
  
  Benchmark.ips do |b|
    b.report("old") { x = BigInt.new(LibGMP::SI::MAX.to_i128 + 1) }
    b.report("new") { x = BigInt.new(num2: LibGMP::SI::MAX.to_i128 + 1) }
  end
  
  Benchmark.ips do |b|
    b.report("old") { x = BigInt.new(LibGMP::UI::MAX.to_i128 + 1) }
    b.report("new") { x = BigInt.new(num2: LibGMP::UI::MAX.to_i128 + 1) }
  end
  
  Benchmark.ips do |b|
    b.report("old") { x = BigInt.new(0x0123456789abcdef_0123456789abcdef_i128) }
    b.report("new") { x = BigInt.new(num2: 0x0123456789abcdef_0123456789abcdef_i128) }
  end
  ```
  ```
  old 117.42M (  8.52ns) (± 1.68%)  16.0B/op   1.03× slower
  new 120.88M (  8.27ns) (± 1.03%)  16.0B/op        fastest
  old   7.21M (138.66ns) (± 1.69%)  80.0B/op  16.99× slower
  new 122.56M (  8.16ns) (± 0.96%)  16.0B/op        fastest
  old   6.52M (153.28ns) (± 0.94%)  80.0B/op  12.75× slower
  new  83.20M ( 12.02ns) (± 0.65%)  32.0B/op        fastest
  old   2.60M (385.20ns) (± 0.34%)  96.0B/op  32.14× slower
  new  83.44M ( 11.98ns) (± 0.66%)  32.0B/op        fastest
  ```

The new undocumented `Int::Primitive#abs_unsigned` method will show up in a few more PRs soon.

### CODE REVIEW

File: src/big/big_int.cr [hunk 1]
```diff
@@ -47,21 +47,40 @@ struct BigInt < Int
...
-  def initialize(num : Int::Signed)
-    if LibC::Long::MIN <= num <= LibC::Long::MAX
-      LibGMP.init_set_si(out @mpz, num)
+  def self.new(num : Int::Primitive)
+    if LibGMP::SI::MIN <= num <= LibGMP::SI::MAX
+      LibGMP.init_set_si(out mpz1, LibGMP::SI.new!(num))
+      new(mpz1)
+    elsif LibGMP::UI::MIN <= num <= LibGMP::UI::MAX
+      LibGMP.init_set_ui(out mpz2, LibGMP::UI.new!(num))
+      new(mpz2)
```
Comment (straight-shoota) [inline src/big/big_int.cr:56]: I'm wondering if it makes sense to reorder the conditions. Assuming that `BigInt` is typically used when operating with numbers that lie outside of UI/SI range, maybe the UI exclusion should come first to have the big numbers outside that range on a fast track? With the current order there are two range checks before reaching the else branch.
```cr
if LibGMP::UI::MIN <= num <= LibGMP::UI::MAX
  if LibGMP::SI::MIN <= num <= LibGMP::SI::MAX
    # init_set_si
  else
    # init_set_ui
  end
else
  # slow path
end
```

**Review by straight-shoota — COMMENTED** [review]
`Int#abs_unsigned` could perhaps be published? I could think it might be useful for user code sometimes...

File: src/big/big_int.cr (hunk 1 above)
Comment (HertzDevil) [reply src/big/big_int.cr:56]: `SI` and `UI`'s ranges do not overlap; the above code would always use the slow path for negative numbers.

Comment (straight-shoota) [reply src/big/big_int.cr:56]: Ah, sorry. I messed up the conditions. This should work:
```cr
if LibGMP::SI::MIN <= num <= LibGMP::UI::MAX
  if num <= LibGMP::SI::MAX
    # init_set_si
  else
    # init_set_ui
  end
else
  # slow path
end
```

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**